### PR TITLE
CR-1119271 ASTeR TC 00.01 - vck5000 - "Device not ready for use" afte…

### PR DESCRIPTION
**Pull request by xdavidz**
…r XRT and platform packages is reloaded

#### Problem solved by the commit
fix loading xsabin metadata from vmr flash, so that even no pkg installed onto the server we still can see the card.

#### How problem was solved, alternative solutions (if any) and why they were rejected
root@xsjyliu51:~# xbmgmt examine
System Configuration
  OS Name              : Linux
  Release              : 5.4.0-96-generic
  Version              : #109-Ubuntu SMP Wed Jan 12 16:49:16 UTC 2022
  Machine              : x86_64
  CPU Cores            : 12
  Memory               : 64109 MB
  Distribution         : Ubuntu 20.04 LTS
  GLIBC                : 2.31
  Model                : PowerEdge T640

XRT
  Version              : 2.13.0
  Branch               : rpu_fpt
  Hash                 : 7d034eed5fb7e9a417c2ed66238845a92146c2b9
  Hash Date            : 2022-01-15 10:48:32
  XOCL                 : 2.13.0, 7d034eed5fb7e9a417c2ed66238845a92146c2b9
  XCLMGMT              : 2.13.0, 7d034eed5fb7e9a417c2ed66238845a92146c2b9

Devices present
  [0000:d9:00.0] : xilinx_vck5000_gen4x8_xdma_base_1 mgmt(inst=55552) 

#### Documentation impact (if any)
https://confluence.xilinx.com/pages/viewpage.action?pageId=302345261#XGQ1.0CommandandQueueSpecification(WIP)-GetLogPage

### CODE REVIEW

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/xgq.c [hunk 1]
```diff
@@ -659,19 +741,105 @@ static int xgq_load_xclbin(struct platform_device *pdev,
...
+			memcpy_from_device(xgq, address, *fw, *fw_size);
+			ret = 0;
+			XGQ_INFO(xgq, "loading fw from vmr size %ld", *fw_size);
+		}
+	}
+done:
+	shm_release_log_page(xgq);
+
+	if (cmd) {
+		remove_xgq_cid(xgq, id);
```
Comment (houlz0507) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/xgq.c:837]: id will be zero if shm_acquire_log_page() fails. Will it be ok to free id 0?

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/xgq.c [hunk 2]
```diff
@@ -659,19 +741,105 @@ static int xgq_load_xclbin(struct platform_device *pdev,
...
+	ret = submit_cmd(xgq, cmd);
+	if (ret) {
+		XGQ_ERR(xgq, "submit cmd failed, cid %d", id);
+		/* return 0, because it is not a firewall trip */
+		ret = 0;
+		goto done;
+	}
+
+	/* wait for command completion */
+	wait_for_completion_interruptible(&cmd->xgq_cmd_complete);
```
Comment (houlz0507) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/xgq.c:809]: should we use wait_for_completion_killable() and check return value?

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/xgq.c [hunk 3]
```diff
@@ -659,19 +741,105 @@ static int xgq_load_xclbin(struct platform_device *pdev,
...
+		vmr_cq_result_copy(xgq, cmd);
+		fw_result = (struct xgq_cmd_cq_log_page_payload *)&xgq->xgq_cq_payload;
+
+		if (fw_result->count > len) {
+			XGQ_ERR(xgq, "need to alloc %d for device data", 
+				fw_result->count);
+			ret = -ENOSPC;
+		} else {
+			*fw_size = fw_result->count;
+			*fw = vmalloc(*fw_size);
```
Comment (houlz0507) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/xgq.c:827]: check *fw == NULL?

Comment (xdavidz) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/xgq.c:833]: thanks, added checking

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/xgq.c (hunk 1 above)
Comment (xdavidz) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/xgq.c]: Yes, I added checking for this. We could refactor the code after this release and move "cmd allocation, cid allocation and shared memory reservation" into one function. But for this release, we may keep it as is for not having potential regression.

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/xgq.c (hunk 2 above)
Comment (xdavidz) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/xgq.c]: yes, fixed

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/xgq.c (hunk 3 above)
Comment (xdavidz) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/xgq.c:833]: added checking

**Review by houlz0507 — APPROVED** [review]

**Review by maxzhen — APPROVED** [review]

**Pull request merged**